### DISCUSSION TITLE
Unable to Implement "App Tracking Transparency" in MAUI for iOS

### DISCUSSION THREAD

**Discussion by barnasoba** [dotnet/maui#17188] [Upgrading from Xamarin]
I'm transitioning an app from Xamarin to MAUI and I'm having difficulty implementing the "App Tracking Transparency" feature, which is essential for iOS apps. I've used the same method as I did in Xamarin, but it doesn't seem to work in MAUI. I've searched extensively, but I can't find documentation or guidance on how to properly incorporate this feature.

Has anyone else faced this issue? Is there a recommended approach or best practices to implement "App Tracking Transparency" in a MAUI app for iOS?

Any help or pointers in the right direction would be greatly appreciated.

Thank you!

**Comment by andrijankara** [comment]
I solve it like this for my app when faced same problem as you:

in Info.plist add key
`<key>NSUserTrackingUsageDescription</key>`
<string>Description</string>

In Maui project root folder add folder interfaces and add interface  IAppTrackingTransparencyService
```
   public interface IAppTrackingTransparencyService
    {
        Task<PermissionStatus> CheckStatusAsync();

        Task RequestAsync(Action<PermissionStatus> completionAction);
    }
```
In Platforms/iOS folder add AppTrackingTransparencyService.cs containing:

```
using AppTrackingTransparency;
using MauiProject.Interfaces;
using static Microsoft.Maui.ApplicationModel.Permissions;

namespace MauiProject.Platforms
{
    public class AppTrackingTransparencyService : BasePlatformPermission,
        IAppTrackingTransparencyService
    {
        protected override Func<IEnumerable<string>> RequiredInfoPlistKeys
            => () => new string[] { "NSUserTrackingUsageDescription" };

        public async Task RequestAsync(Action<PermissionStatus> completionAction)
        {
            // Due to some changes on iOS 15 we need to wait a little bit before using
            // ATTrackingManager to show the corresponding popup
            await Task.Delay(1000);

            ATTrackingManager.RequestTrackingAuthorization(
                (result) => completionAction(ConvertStatus(result)));
        }

        public override Task<PermissionStatus> CheckStatusAsync()
        {
            return Task.FromResult(
                ConvertStatus(ATTrackingManager.TrackingAuthorizationStatus));
        }

        private PermissionStatus ConvertStatus(
            ATTrackingManagerAuthorizationStatus status)
        {
            switch (status)
            {
                case ATTrackingManagerAuthorizationStatus.NotDetermined:
                    return PermissionStatus.Disabled;
                case ATTrackingManagerAuthorizationStatus.Restricted:
                    return PermissionStatus.Restricted;
                case ATTrackingManagerAuthorizationStatus.Denied:
                    return PermissionStatus.Denied;
                case ATTrackingManagerAuthorizationStatus.Authorized:
                    return PermissionStatus.Granted;
                default:
                    return PermissionStatus.Unknown;
            }
        }
    }
}
```
In MauiProgram.cs register interface only if device is iOS :
```
#if IOS
        builder.Services.AddTransient<IAppTrackingTransparencyService, AppTrackingTransparencyService>();

```
In App.xaml.cs inject created transparency tracking service ( if needed use pretprocessor directives "if IOS")
```
IAppTrackingTransparencyService appTrackingTransparencyService;

public App(IAppTrackingTransparencyService _trackingService)
{
    InitializeComponent();
    App.Current.UserAppTheme = AppTheme.Light;
    appTrackingTransparencyService = _trackingService;
        MainPage = new AppShell();
}
```
in same file override OnStart() method ( if needed use pretprocessor directives "if IOS")
```
    protected override async void OnStart()
    {
#if IOS
        if (DeviceInfo.Platform == DevicePlatform.iOS)
        {
            var appTrackingTransparencyService = DependencyService.
                       Get<IAppTrackingTransparencyService>();
            try
            {
                var status = await
                      appTrackingTransparencyService.CheckStatusAsync();

                if (status != PermissionStatus.Granted)
                    await appTrackingTransparencyService.
                                RequestAsync(s => HandleTracking(s));
                else
                    HandleTracking(status);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                HandleTracking(null);
            }
           
        }
#endif
    }
    private void HandleTracking(PermissionStatus? status)
    {
        if (status != PermissionStatus.Granted)
            return;

        // enable tracking
    }
```

**Reply by barnasoba** [reply]
Thank you very much @andrijankara , I have tried it but I can't get it to work. Do you have a repository where I can consult it?

**Reply by ozersenol** [reply]
Me neither I could not make it work and prompt for permission. Any help please.

**Comment by drasticactions** [comment]
https://github.com/drasticactions/MauiRepros/blob/main/AppTrackingTransparencyMaui/Platforms/iOS/AppTrackingTransparencyService.cs

https://github.com/drasticactions/MauiRepros/blob/main/AppTrackingTransparencyMaui/Platforms/iOS/Info.plist#L31-L32

You don't need the Dependency Injection service if you don't want it. I think the missing part was not using the `ATTrackingManager.RequestTrackingAuthorizationAsync` and relying on the callback method. 

<img width="423" alt="スクリーンショット 2023-11-27 15 43 29" src="https://github.com/dotnet/maui/assets/898335/3ef36ae2-1dc0-4dd8-a910-9877b6949f73">
<img width="386" alt="スクリーンショット 2023-11-27 15 43 39" src="https://github.com/dotnet/maui/assets/898335/55fd4ad9-5290-41f3-853b-48aa3a99ac8c">

Doing this, it should function correctly.
